src/MatrixTwitter.js

```diff
@@ -0,0 +1,1040 @@
+var Twitter = require('twitter');
+var Request = require('request');
+var fs = require('fs');
+var log = require('npmlog');
+var Buffer = require('buffer').Buffer;
+var HTMLDecoder = new require('html-entities').AllHtmlEntities;
+var MatrixRoom = require("matrix-appservice-bridge").MatrixRoom;
+var RemoteRoom = require("matrix-appservice-bridge").RemoteRoom;
+
+var ProcessedTweetList = require("./ProcessedTweetList.js");
+var util = require("./util.js");
+
+const TWITTER_CLIENT_INTERVAL_MS    = 60000;
+const TWITTER_MSG_QUEUE_INTERVAL_MS = 1500;
+const TIMELINE_POLL_INTERVAL = 3050; //Twitter allows 300 calls per 15 minute (We add 50 milliseconds for a little safety).
+const HASHTAG_POLL_INTERVAL = 3050; //Twitter allows 450 calls per 15 minute (We add 50 milliseconds for a little safety).
+const TWEET_REPLY_MAX_DEPTH = 3;
+
+/**
+ * MatrixTwitter - This class handles the connections between the Twitter API
+ * and the bridge.
+ * @class
+ * @param  {matrix-appservice-bridge.Bridge} bridge
+ * @param  config. Bridge configuration (currently only make use of the auth
+*  information)
+ * @param  config.app_auth.consumer_key Twitter consumer key
+ * @param  config.app_auth.consumer_secret Twitter consumer secret
+ * @param  {TwitterDB}       storage
+ */
+var MatrixTwitter = function (bridge, config, storage) {
+  this.app_auth = config.app_auth;
+  this.app_twitter = null;
+  this.tclients = new Map(); // {'@userid':TwitterClient}
+  this.storage = storage;
+
+  this.timeline_queue = [];
+  this.timeline_intervalID = null;
+
+  this.hashtag_intervalID = null;
+  this.hashtag_queue = [];
+
+  this.user_streams = new Map(); // {'@userid':TwitterStream}
+
+  this.processed_tweets = new ProcessedTweetList(256, 32);  //This will contain all the tweet IDs of things we don't want to repeat.
+  this.sent_dms = new ProcessedTweetList(1, 1);  //This will contain the body of the DM posted to a room to avoid reposting it.
+  this.msg_queue = [];
+  this.msg_queue_intervalID = null;
+  this._bridge = bridge;
+};
+
+
+/**
+ * MatrixTwitter.prototype.start - Starts the timers for polling the API and
+ * authenticates the application with Twitter.
+ * @function
+ * @return {Promise}  A promise that returns when authentication succeeds.
+ */
+MatrixTwitter.prototype.start = function () {
+  if(this.app_twitter != null) {
+    log.warn("Twitter",  "Attempted to call start() while having been started previously.");
+    return
```

Comment:
You can race here if you call `start()` twice in a row because you don't immediately non-nullify `this.app_twitter` on this turn of the event loop. If you guard using the promise itself, you'll actually protect yourself from calling `start()` twice.